### PR TITLE
prov/efa: eliminate the recv_wr in packet entry and sendv_pool

### PR DESCRIPTION
**Pull request by wzamazon**
a series of patches that eliminates the recv_wr in packet entry and the sendv_pool, and clean up space in `efa_rdm_pke`

### CODE REVIEW

**Comment by wzamazon** [comment]
First 3 commits are from https://github.com/ofiwg/libfabric/pull/9058. Please focus on the last 5

File: prov/efa/src/rdm/efa_rdm_pke.h [hunk 1]
```diff
@@ -207,30 +179,56 @@ struct efa_rdm_pke {
...
+	 * @brief payload buffer
+	 * 
+	 * @details
+	 * "payload" refers to a piece of user data that is
+	 * going over wire.
+	 *
+	 * the pointer "payload" can be directly pointing to
+	 * user's buffer or provider's internal bounce
+	 * buffer (where user data has been copied to).
 	 */
```
Comment (wenduwan) [inline prov/efa/src/rdm/efa_rdm_pke.h:200]: TBH this makes payload even more confusing.... do you mean it is actually `localdata`?

payload typically means the same thing as `wiredata`, which is the actual data sent over the network.

File: prov/efa/test/efa_unit_test_cq.c [hunk 2]
```diff
@@ -74,111 +74,88 @@ void test_ibv_cq_ex_read_empty_cq(struct efa_resource **state)
...
-
-		efa_rdm_ep->host_id = local_host_id;
-
-		ibv_qp =  efa_rdm_ep->base_ep.qp->ibv_qp;
+	efa_rdm_ep = container_of(resource->ep, struct efa_rdm_ep, base_ep.util_ep.ep_fid);
+	efa_rdm_ep->host_id = local_host_id;
+	ibv_qpx = efa_rdm_ep->base_ep.qp->ibv_qp_ex;
+	ibv_cqx = efa_rdm_ep->ibv_cq_ex;
+	/* set use_shm_for_tx to false to force efa_rdm_ep to use efa device to send,
+	 * which means call ibv_post_send
```
Comment (wenduwan) [inline prov/efa/test/efa_unit_test_cq.c:109]: I searched `ibv_post_send` on this page there are a few places. I think you intended to remove them.

**Comment by wenduwan** [comment]
CI failure
```
11:32:48 prov/efa/src/rdm/efa_rdm_ep_progress.c: In function 'efa_rdm_ep_bulk_post_internal_rx_pkts':
11:32:48 prov/efa/src/rdm/efa_rdm_ep_progress.c:86:18: error: 'nrecv' undeclared (first use in this function); did you mean 'recv'?
11:32:48   for (i = 0; i < nrecv; ++i) {
11:32:48                   ^~~~~
11:32:48                   recv
11:32:48 prov/efa/src/rdm/efa_rdm_ep_progress.c:86:18: note: each undeclared identifier is reported only once for each function it appears in
11:32:48 make[1]: *** [prov/efa/src/rdm/src_libfabric_la-efa_rdm_ep_progress.lo] Error 1
11:32:48 make[1]: *** Waiting for unfinished jobs....
11:32:48 make: *** [all] Error 2
```

File: prov/efa/src/rdm/efa_rdm_pke.h (hunk 1 above)
Comment (wzamazon) [reply prov/efa/src/rdm/efa_rdm_pke.h:200]: > payload typically means the same thing as wiredata, which is the actual data sent over the network.

Yes. "payload" is the actual data that is going over wire.

EFA device support 2 iov, which means we can send 2 buffers that are not continuous in 1 packet.

wiredata consume 1st iov, which always have the header.

payload consumes 2nd iov, which has the user's data. Sometimes it is pointing directly to user's buffer.

What kind of wording would you suggest to make it more clear?

File: prov/efa/test/efa_unit_test_cq.c (hunk 2 above)
Comment (wzamazon) [reply prov/efa/test/efa_unit_test_cq.c:109]: I removed the reference to `ibv_post_send` in the other PR https://github.com/ofiwg/libfabric/pull/9058. Will rebase this PR once that one is merged

**Comment by wzamazon** [comment]
> CI failure

Fixed

File: prov/efa/src/rdm/efa_rdm_pke.h (hunk 1 above)
Comment (wzamazon) [reply prov/efa/src/rdm/efa_rdm_pke.h]: I updated the comment after our offline discussion. Please take a look

**Comment by wzamazon** [comment]
> CI failure

CI failure has been fixed

**Review by wenduwan — APPROVED** [review]

**Pull request merged**